[WebCore] Update Bison version for parser to 3.8.2

LumioseSil opened the pull request:
<pre>[WebCore] Update Bison version for parser to 3.8.2
https://bugs.webkit.org/show_bug.cgi?id=278609

Reviewed by NOBODY (OOPS!).

Code is currently generated using Bison 2.3.
We can instead use Bison 3.8.2, as it is more modern,
and has a legal exception to it's license so we can use it.
(See: https://www.gnu.org/software/bison/manual/html_node/Conditions.html)

* Source/WebCore/xml/XPathGrammar.cpp: Regenerate.
* Source/WebCore/xml/XPathGrammar.h: Regenerate.
* Source/WebCore/xml/XPathGrammar.y: Replace "%pure-parser" with
  "%define api.pure full."
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/125f973f34b3516a94e05a12fbb3c690704bc5da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64350 "477 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16947 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14958 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15238 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10320 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40405 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13056 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13832 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59041 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13386 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12904 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8331 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55768 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6858 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/544 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39527 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40349 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/xml/XPathGrammar.h:
```diff
@@ -1,26 +1,22 @@
    it under the terms of the GNU General Public License as published by
-   the Free Software Foundation; either version 2, or (at your option)
-   any later version.
+   the Free Software Foundation, either version 3 of the License, or
```
We can't accept this license change.  That's why we're stuck at an old version.

aperezdc replied:
Bison contains a [licensing exception](https://www.gnu.org/software/bison/manual/html_node/Conditions.html) which allows to use its output under any other license, and the exception is itself reproduced in the output files ([implementation](https://github.com/WebKit/WebKit/pull/32672/files#diff-f30aed71e815bddcdab2f2b46e619f5b7980db0ae6d915a64c0885acef25d3a4R25), [header](https://github.com/WebKit/WebKit/pull/32672/files#diff-3006c592ec6e6844ebcc968f77092eb16f1b911051e99bf3ff3c4ed38e91fbc6R21)).

IANAL, but the text is crystal clear in this regard.

AZero13 replied:
> We can't accept this license change. That's why we're stuck at an old version.

That would be true if we were shipping Bison itself. We aren't. 

Let me obtain evidence to prove it since I know this is a big deal.

AZero13 replied:
/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.

   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

AZero13 replied:
Oh, and by the way, if this is not clear enough then there is a problem because it turns out: ANGLE uses code generated from Bison 3.8, which means that either ANGLE cannot use the code, or we can as well.
<img width="711" alt="Screenshot 2024-08-27 at 8 43 40 AM" src="https://github.com/user-attachments/assets/4e5b42dc-443f-4452-a507-726bcfca7d01">

LumioseSil committed:
[WebCore] Update Bison version for parser to 3.8.2
https://bugs.webkit.org/show_bug.cgi?id=278609

Reviewed by NOBODY (OOPS!).

Code is currently generated using Bison 2.3.
We can instead use Bison 3.8.2, as it is more modern,
and has a legal exception to it's license so we can use it.
(See: https://www.gnu.org/software/bison/manual/html_node/Conditions.html)

* Source/WebCore/xml/XPathGrammar.cpp: Regenerate.
* Source/WebCore/xml/XPathGrammar.h: Regenerate.
* Source/WebCore/xml/XPathGrammar.y: Replace "%pure-parser" with
  "%define api.pure full."

achristensen07 reviewed (changes requested):
This is not acceptable as it now stands.  It needs to at least have this comment like the other content generated by Bison 3.8.2:

/* Apple Note: For the avoidance of doubt, Apple elects to distribute this file under the terms of
 * the BSD license. */

We will also have to do an internal legal review to get permission to have such code with such a clarification.  Thereafter, we would probably be unable to modify and regenerate the code without help from Igalia or others because macOS ships with Bison 2.3.  This is a significant hinderance, which is why we have intentionally chosen not to do this in the past.  If you can point to a specific benefit for a specific case where the update would make a specific difference, we may consider going through that process.  It is currently unclear what you mean by "uses K&R for some functions" and what the binary output is before and after this change and how this change is helpful.  It seems like you may be trying to avoid a warning in some future version of some compiler, but I'm not sure.
